src/documentation/io/testing.md

```diff
@@ -108,11 +108,464 @@ This allows the I/O transform's unit tests to act as if they have a perfect netw
+    *   The only verification of the data should be the result from the read. Don't validate the data written to the database in any other way.
+    *   Validate the actual contents of all rows in an efficient manner. An easy way to do this is by taking a hash of the rows and combining them. `HashingFn` can help make this simple, and `TestRow` has pre-computed hashes.
+    *   For easy debugging, use `PAssert`'s `containsInAnyOrder` to validate the contents of a subset of all rows.
+*   **Tests should assume they may be run multiple times and/or simultaneously on the same database instance.**
+    *   Clean up test data: do this in an `@afterClass` to ensure it runs.
+    *   Use unique table names per run (timestamps are an easy way to do this) and per-method where appropriate.
+
+An end to end example of these principles can be found in [JdbcIOIT](https://github.com/ssisk/beam/blob/jdbc-it-perf/sdks/java/io/jdbc/src/test/java/org/apache/beam/sdk/io/jdbc/JdbcIOIT.java).
+
+
+#### Kubernetes scripts {#kubernetes-scripts}
+
+As discussed in [Integration tests, data stores, and Kubernetes](#integration-tests-data-stores-and-kubernetes), to have your tests run on Beam's continuous integration server, you'll need to implement a Kubernetes script that creates an instance of your data store.
+
+If you would like help with this or have other questions, contact the Beam dev@ mailing list and the community may be able to assist you.
+
+Guidelines for creating a Beam data store Kubernetes script:
+1.  **You must only provide access to the data store instance via a `NodePort` service.**
+    *   This is a requirement for security, since it means that only the local network has access to the data store. This is particularly important since many data stores don't have security on by default, and even if they do, their passwords will be checked in to our public Github repo.
+1.  **Convention: define two Kubernetes scripts.**
+    *   This is the best known way to implement item #1.
+    *   The first script will be the core script run on the Beam CI servers and should contain the main datastore instance script (`StatefulSet`) plus a `NodePort` service exposing the data store.
+    *   The second script will define a `LoadBalancer` service, used for local development if the Kubernetes cluster is on another network. This file's name is usually suffixed with '-for-local-dev'.
+1.  **You must ensure that pods are recreated after crashes.**
+    *   If you use a `pod` directly, it will not be recreated if the pod crashes or something causes the cluster to move the container for your pod.
+    *   In most cases, you'll want to use `StatefulSet` as it supports persistent disks that last between restarts, and having a stable network identifier associated with the pod using a particular persistent disk. `Deployment`s/`ReplicaSet`s are also possibly useful, but likely in fewer scenarios since they do not have those features.
```

Comment:
The 'Deployment's/'ReplicaSet's looks kinda weird in the output html because of the space between the term and the 's'